### PR TITLE
Voyager unit map considering linear sequence of levels

### PR DESCRIPTION
**Pull request by shubhi1092**
- Added nextLevels and first property to campaign and classroom schema
- Updated next level logic for voyager classrooms and campaigns
- New URL for voyager unit map
- New file for voyager utils
- Remove 'intro' level type from the level schema (will handle it as part of intro levels implementation)

Updating the next level logic impacts some existing code as well, but it wont affect any production users until they have a voyager course in their classrooms (new logic will be used only for voyager courses). Code paths affected due to this:
- Calculating next level in course victory modal
- Calculating next level in classroom stats 

Also, added some TODO comments about edge cases, and future work.

NOTE: The focus of this PR is just the functionality, and not UI. It will depend largely on the UI specs, hence using very basic and minimal UI for now.

How to test:
1. Use the steps in the script `voyagerWorldMapData` on the server side PR to test the home version.
2. To test the classroom version, you will need to create a new course in your local db using any voyager campaign and then assign it to a student in the classroom. Please contact me for any questions related to this.

~TODO: Tests coming soon!~

Related server PR: https://github.com/codecombat/codecombat-server/pull/73

### CODE REVIEW

File: app/core/Router.coffee [hunk 1]
```diff
@@ -196,6 +196,12 @@ module.exports = class CocoRouter extends Backbone.Router
       @navigate("play/web-dev-level/#{sessionID}?#{queryString}", { trigger: true, replace: true })
     'play/spectate/:levelID': go('play/SpectateView')
     'play/:map': go('play/CampaignView')
+    'play-voyager/:unit(?course-instance=:courseInstanceId)': (unit, courseInstanceId) ->
```
Comment (AndrewJakubowicz) [inline app/core/Router.coffee]: 👏 Love seeing the new vue routes 👏

File: app/core/utils.coffee [hunk 2]
```diff
@@ -119,6 +119,11 @@ courseIDs =
   COMPUTER_SCIENCE_5: '569ed916efa72b0ced971447'
   COMPUTER_SCIENCE_6: '5817d673e85d1220db624ca4'
 
+# TODO add when final courses content created for voyager 
+voyagerCourseIDs = [
+  ''
```
Comment (AndrewJakubowicz) [inline app/core/utils.coffee]: Could this empty string create issues, and could an empty array work?

File: app/core/voyagerUtils.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
+
+module.exports = {
+
+  /**
+   * Creates a level status map using the list of level session objects.
+   * @param {Object[]} sessions - The list of level session objects.
+   * @returns {Object} - Object with key as the level original id, and value as complete/started.
+   */
+  getLevelStatusMap: function (sessions) {
+    let levelStatusMap = {}
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: As this isn't being rebound, should we prefer `const`?
```suggestion
    const levelStatusMap = {}
```

File: app/core/voyagerUtils.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    for (let session of sessions) {
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: Nitpick: Similar comment about `const`. [Reference for my reasoning](https://medium.com/tanda-developers/why-do-we-prefer-const-d805b9dc03be)

File: app/core/voyagerUtils.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    return levelDataMap
+  },
+
+  /**
+   * Gets the next level original id for a given level
+   * @param {Object} level - The level object.
+   * @param {Object[]} level.nextLevels - The array of nextLevels for the given level.
+   * @returns {string|undefined} - 'Original' id of the next level, or undefined.
+   */
+  getNextLevelOriginalForLevel: function (level) {
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: I agree there is currently only one level currently. However once there are multiple levels we will want to extend this function.
Should we instead return an array of strings as possible next levels?
Then when there are many next levels this logic won't change.

File: app/core/voyagerUtils.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      return nextLevels[0].levelOriginal // assuming that there will be just one next level for voyager v1 as of now.
+    }
+    return undefined
+  },
+
+  /**
+   * Calculates all the next levels for a list of levels in a classroom/campaign and level sessions.
+   * @param {Object[]} sessions - The list of level session objects.
+   * @param {Object[]|Object} levels - The list of level objects, or an object with keys as level original id and value as level data.
+   * @param {Object[]} levels.nextLevels - The array of nextLevels for a level.
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: Minor JsDoc nitpick. `@param` signifies a function argument.
If you want to describe the type of `Object` for `levels`, you can use [`@typedef`](http://usejsdoc.org/tags-typedef.html).

File: app/core/voyagerUtils.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      levelStatusMap = this.getLevelStatusMap(sessions)
+    }
+    let nextLevels = [] // next level = started levels + incomplete unlocked levels + not-started first levels
+    let levelDataMap = {}
+    if (_.isArray(levels)) {
+      levelDataMap = this.getLevelDataMap(levels)
+    } else {
+      levelDataMap = levels
+    }
+    for (let [i, level] of Object.entries(levelDataMap)) {
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: Should `i` have a more descriptive name?

File: app/core/voyagerUtils.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      if (levelStatusMap[i] === 'started' && nextLevels.indexOf(level) < 0) {
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: Does `indexOf` work with the level object?

File: app/core/voyagerUtils.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        nextLevels.push(level)
+      } else if (levelStatusMap[i] === 'complete') {
+        let nextLevelOriginal = this.getNextLevelOriginalForLevel(level)
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: Should this be extended to handle multiple next levels? It should be a minor extension.

File: app/core/voyagerUtils.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+   * There would only be 1 next level for voyager v1 as of now, hence use the first element of the array returned from here.
+   * TODO: Edge case - If the 'sessions' contain a session for a level which has been played in another classroom, or another campaign;
+   * then it will be part of the nextLevels even if not unlocked for the current classroom/campaign.
+  */
+  findNextLevels: function (sessions, levels, levelStatusMap) {
+    if (!levelStatusMap) {
+      levelStatusMap = this.getLevelStatusMap(sessions)
+    }
+    let nextLevels = [] // next level = started levels + incomplete unlocked levels + not-started first levels
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: Much of the checking logic could potentially be removed  if this was a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) of nextLevel strings.

Then you could just use [`nextLevels.set(levelOriginalStringId)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/add) and it would automatically only keep unique levels.
The set can be changed into an array and returned at the end  of the function.
This will make it much easier to test this specific, and very critical function.
An array of level strings is easier to test  than an array of level objects.
Then we could use a different function call outside this function to transform the level string original  ids back into level  objects.

File: app/models/Classroom.coffee [hunk 11]
```diff
@@ -167,10 +169,15 @@ module.exports = class Classroom extends CocoModel
...
       needsPractice = utils.needsPractice(currentPlaytime, currentLevel.get('practiceThresholdMinutes')) and not currentLevel.get('assessment')
-      nextIndex = utils.findNextLevel(levels, currentIndex, needsPractice)
-    nextLevel = courseLevels.models[nextIndex]
-    nextLevel = arena if levelsLeft is 0
-    nextLevel ?= _.find courseLevels.models, (level) -> not levelSessionMap[level.get('original')]?.get('state')?.complete
+      unless utils.voyagerCourseIDs.includes(courseID)
+        nextIndex = utils.findNextLevel(levels, currentIndex, needsPractice)
+    if utils.voyagerCourseIDs.includes(courseID)
+      # assuming that there will be only one next level in voyager v1 for now
+      nextLevel = new Level(voyagerUtils.findNextLevels(sessions, courseLevels.models)[0])
```
Comment (AndrewJakubowicz) [inline app/models/Classroom.coffee]: In  previous comments I have mentioned the single next level assumption. That is due to it being new code that we want to write realistic tests for. If we lean too heavily into our nextLevel assumption. The tests will end up reflecting that and it won't be easy  to change.

However. This code is now interacting with the existing legacy system. I think it's excellent to leave a comment here about the assumption and only assign the one next level.
:clap:

File: app/schemas/models/level.coffee [hunk 12]
```diff
@@ -264,7 +264,7 @@ _.extend LevelSchema.properties,
   i18n: {type: 'object', format: 'i18n', props: ['name', 'description', 'loadingTip', 'studentPlayInstructions'], description: 'Help translate this level'}
   banner: {type: 'string', format: 'image-file', title: 'Banner'}
   goals: c.array {title: 'Goals', description: 'An array of goals which are visible to the player and can trigger scripts.'}, GoalSchema
-  type: c.shortString(title: 'Type', description: 'What type of level this is.', 'enum': ['campaign', 'ladder', 'ladder-tutorial', 'hero', 'hero-ladder', 'hero-coop', 'course', 'course-ladder', 'game-dev', 'web-dev', 'intro'])
+  type: c.shortString(title: 'Type', description: 'What type of level this is.', 'enum': ['campaign', 'ladder', 'ladder-tutorial', 'hero', 'hero-ladder', 'hero-coop', 'course', 'course-ladder', 'game-dev', 'web-dev'])
```
Comment (AndrewJakubowicz) [inline app/schemas/models/level.coffee]: Great cleanup. :clap: Agree with your reasoning :)

File: app/views/play/VoyagerUnitMap.vue [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+          const expectedLanguage = classroomLevel.primerLanguage || this.classroom.aceConfig.language
+          if (session.codeLanguage != expectedLanguage) {
+            this.levelSessions.splice(this.levelSessions.indexOf(session),1)
+          } 
+        }
+      }
+      this.levelStatusMap = utils.getLevelStatusMap(this.levelSessions)
+    },
+    determineNextLevel() { // set .next and .locked for this.levels
+      debugger
```
Comment (AndrewJakubowicz) [inline app/views/play/VoyagerUnitMap.vue]: typo?

File: app/views/play/VoyagerUnitMap.vue [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+            continue
+          const expectedLanguage = classroomLevel.primerLanguage || this.classroom.aceConfig.language
+          if (session.codeLanguage != expectedLanguage) {
+            this.levelSessions.splice(this.levelSessions.indexOf(session),1)
+          } 
+        }
+      }
+      this.levelStatusMap = utils.getLevelStatusMap(this.levelSessions)
+    },
+    determineNextLevel() { // set .next and .locked for this.levels
```
Comment (AndrewJakubowicz) [inline app/views/play/VoyagerUnitMap.vue]: I'm uncomfortable with complex logic in the Vue components. Should logic be written in a helper or related file and imported.
Testing logic + view rendering compounds complexity.
If we separate out logic we  can test that separate, and then test the vue component more easily by passing in the correct props. 
I do agree that some logic can't be avoided. But these quickly complicated methods like `determineNextLevel` feel like they should stand alone and be tested alone.

File: app/views/play/level/modal/CourseVictoryModal.coffee [hunk 15]
```diff
@@ -36,15 +37,17 @@ module.exports = class CourseVictoryModal extends ModalView
...
+    unless utils.voyagerCourseIDs.includes(@courseID)
+      nextLevelPromise = api.levels.fetchNextForCourse({
+        levelOriginalID: @level.get('original')
+        @courseInstanceID
+        @courseID
+        sessionID: @session.id
+      }).then ({ level, assessment }) =>
+        @nextLevel.set(level)
+        @nextAssessment.set(assessment)
+      @supermodel.trackPromise(nextLevelPromise)
```
Comment (AndrewJakubowicz) [inline app/views/play/level/modal/CourseVictoryModal.coffee]: Interested in learning about what `trackPromise`  method is adding to this promise?

File: app/views/play/level/modal/CourseVictoryModal.coffee [hunk 16]
```diff
@@ -98,6 +101,16 @@ module.exports = class CourseVictoryModal extends ModalView
         rewardsView.on 'continue', @onViewContinue, @
         @views.push(rewardsView)
 
+    # get next level for voyager course, no `nextAssessment` in voyager
+    if utils.voyagerCourseIDs.includes(@courseID) and @classroom
+      currentLevel = @classroom.get('courses')?.find((c) => c._id == @courseID)?.levels?.find((l) => l.original == @level.get('original'))
+      nextLevelOriginal = voyagerUtils.getNextLevelOriginalForLevel(currentLevel)
```
Comment (AndrewJakubowicz) [inline app/views/play/level/modal/CourseVictoryModal.coffee]: Similar thoughts about one vs many next levels.

File: app/core/utils.coffee (hunk 2 above)
Comment (shubhi1092) [reply app/core/utils.coffee]: Yeah I think an empty array would work, will update. Thank you

File: app/views/play/VoyagerUnitMap.vue (hunk 13 above)
Comment (shubhi1092) [reply app/views/play/VoyagerUnitMap.vue]: thanks for catching that

File: app/views/play/level/modal/CourseVictoryModal.coffee (hunk 16 above)
Comment (shubhi1092) [reply app/views/play/level/modal/CourseVictoryModal.coffee]: Not yet, in the future if we have multiple next levels, the logic will depend on how do we want the game flow to be.
But I would want to handle all that complicated logic in `getNextLevelOriginalForLevel` as much as possible, and just return whats relevant to course victory modal.

File: app/views/play/level/modal/CourseVictoryModal.coffee (hunk 15 above)
Comment (shubhi1092) [reply app/views/play/level/modal/CourseVictoryModal.coffee]: It mostly adds the callbacks for success and failure to the promise and handles the loading of the page based on that. I don't know exactly how, but it calls `onLoaded` once all the tracked requests and promises are completed.

File: app/core/voyagerUtils.js (hunk 4 above)
Comment (shubhi1092) [reply app/core/voyagerUtils.js]: Thank you for adding that link, it was useful.

File: app/core/voyagerUtils.js (hunk 7 above)
Comment (shubhi1092) [reply app/core/voyagerUtils.js]: Updating it to `levelOriginal`

File: app/core/voyagerUtils.js (hunk 5 above)
Comment (shubhi1092) [reply app/core/voyagerUtils.js]: Yeah that's a good point, will update. Thank you

File: app/views/play/level/modal/CourseVictoryModal.coffee (hunk 15 above)
Comment (AndrewJakubowicz) [reply app/views/play/level/modal/CourseVictoryModal.coffee]: Thanks for explaining :)

File: app/core/voyagerUtils.js (hunk 10 above)
Comment (shubhi1092) [reply app/core/voyagerUtils.js]: Using a `set` is a great idea. Also thanks for the suggestion of making it simpler by returning an array of strings instead of objects. :)

File: app/core/voyagerUtils.js (hunk 8 above)
Comment (shubhi1092) [reply app/core/voyagerUtils.js]: Updating to use `set` instead.

File: app/core/voyagerUtils.js (hunk 6 above)
Comment (shubhi1092) [reply app/core/voyagerUtils.js]: I referred to this section(http://usejsdoc.org/tags-param.html#parameters-with-properties) for documenting some properties that are expected in the `levels object`.
I am not sure if we should use typedef here since it is about documenting `custom type`.

File: app/views/play/VoyagerUnitMap.vue (hunk 14 above)
Comment (shubhi1092) [reply app/views/play/VoyagerUnitMap.vue]: That's a good idea, although that would mean moving all the logic from the `methods` section to some helper/utility file. I am not sure when to do that, vs have a method in the vue file itself.
I believe it's about how we want to set the best practices for vue going forward, and we can bring up this discussion in the vue best practices doc(if it doesnt include it already).

Although I like the idea, I am keeping as it is, for now, and will update later based on the vue best practices we establish.

Comment (shubhi1092) [reply app/views/play/VoyagerUnitMap.vue]: Also, on that note, I tried to have some common pieces like  `getLevelStatusMap`, and `findNextLevels` in a common utility file called `voyagerUtils.js`.

File: app/core/voyagerUtils.js (hunk 6 above)
Comment (AndrewJakubowicz) [reply app/core/voyagerUtils.js]: I'm completely wrong! This solves a major pain point for me in my documenting!!!
Thanks for the reference!

File: app/core/voyagerUtils.js [hunk 17]
```diff
@@ -38,56 +38,77 @@ module.exports = {
...
-   * @returns {string|undefined} - 'Original' id of the next level, or undefined.
+   * @returns {string[]]} - Array of next level 'Original' ids.
    */
   getNextLevelOriginalForLevel: function (level) {
-    let nextLevels = level.nextLevels
+    const nextLevels = level.nextLevels
+    const nextLevelOriginals = []
     if ((nextLevels || []).length > 0) {
-      return nextLevels[0].levelOriginal // assuming that there will be just one next level for voyager v1 as of now.
+      nextLevelOriginals.push(nextLevels[0].levelOriginal) // assuming that there will be just one next level for voyager v1 as of now.
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: Fine to leave as is. Mostly asking for my understanding. As we're getting the next level original of only the first level it feels like the code could avoid the conditional and nextLevelOriginals resulting array. It could also be a map over the nextLevels, returning the originals.

Something like:

```
(level.nextLevels || [])
  .map(level => level.levelOriginal)
```

Although from your comment the design appears intentional. Is there any additional details regarding how this method will change when the 1FH capstone level is added?

File: app/core/voyagerUtils.js [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+      levelDataMap = this.getLevelDataMap(levels)
+    } else {
+      levelDataMap = levels
+    }
+    for (let [levelOriginal, level] of Object.entries(levelDataMap)) {
+      if (levelStatusMap[levelOriginal] === 'started') {
+        nextLevelOriginals.add(levelOriginal)
+      } else if (levelStatusMap[levelOriginal] === 'complete') {
+        const unlockedLevelOriginals = this.getNextLevelOriginalForLevel(level)
+        if (unlockedLevelOriginals.length > 0) { // incomplete unlocks
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: This check may not be required. As long as `this.getNextLevelOriginalForLevel(level)` always returns an array. Or by adding an `|| []` after it. The `filter` and `forEach` will do nothing with an empty array.

File: app/core/voyagerUtils.js [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+          unlockedLevelOriginals
+            .filter((original) => levelStatusMap[original] !== 'complete')
+            .forEach((original) => nextLevelOriginals.add(original))
+        }
+      } else if (level.first) {
+        nextLevelOriginals.add(levelOriginal)
+      }
+    }
+    return [...nextLevelOriginals]
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: 👏 Didn't know you could do this. 👏 Excellent!

File: app/core/voyagerUtils.js [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    const nextLevelOriginals = new Set() // next level = started levels + incomplete unlocked levels + not-started first levels
+
+    let levelDataMap = {}
+    if (_.isArray(levels)) {
+      levelDataMap = this.getLevelDataMap(levels)
+    } else {
+      levelDataMap = levels
+    }
+    for (let [levelOriginal, level] of Object.entries(levelDataMap)) {
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: Nitpick: `const`

File: app/core/voyagerUtils.js [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+    return levelStatusMap
+  },
+
+  /**
+   * Creates a level data map using the list of level objects.
+   * @param {Object[]} levels - The list of level objects.
+   * @returns {Object} - Object with key as the level original id, and value as the object containing level's data.
+   */
+  getLevelDataMap: function (levels) {
+    const levelDataMap = {}
```
Comment (AndrewJakubowicz) [inline app/core/voyagerUtils.js]: Regarding the name of the method, is there a rationale for returning an Object literal vs a [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)?

Should the method be named something like `getLevelData` to not suggest the Map data structure?

File: app/lib/dynamicRequire.js [hunk 22]
```diff
@@ -87,6 +87,7 @@ module.exports = {
   'views/user/IdentifyView': function () { return import(/* webpackChunkName: "user" */ 'views/user/IdentifyView') },
   'views/account/IsraelSignupView': function () { return import(/* webpackChunkName: "account" */ 'views/account/IsraelSignupView') },
   'views/LegalView': function () { return import(/* webpackChunkName: "LegalView" */ 'views/LegalView') },
+  'views/play/VoyagerUnitMap': function() { return import(/*webpackChunkName: "play" */ 'views/play/VoyagerUnitMap') },
```
Comment (AndrewJakubowicz) [inline app/lib/dynamicRequire.js]: Should we, and could we start a separate ozaria webpack chunk for the new ozaria pieces?

Thus as we make infrastructure and front end improvements for ozaria. We won't be blowing up the size of the play chunk for the code combat players.
And we might be able to make the ozaria chunk smaller as well?

File: app/schemas/models/campaign.schema.coffee [hunk 23]
```diff
@@ -78,6 +78,20 @@ _.extend CampaignSchema.properties, {
       #- normal properties
       position: c.point2d()
 
+      # properties relevant for voyager campaigns
+      nextLevels: {   # array of next levels
+        type: 'array'
+        format: 'next-level'
```
Comment (AndrewJakubowicz) [inline app/schemas/models/campaign.schema.coffee]: What does this format field do?

File: app/schemas/models/classroom.schema.coffee [hunk 24]
```diff
@@ -39,6 +39,19 @@ _.extend ClassroomSchema.properties,
       name: {type: 'string'}
       slug: {type: 'string'}
       position: c.point2d()
+
+      # properties relevant for voyager campaigns
```
Comment (AndrewJakubowicz) [inline app/schemas/models/classroom.schema.coffee]: Should this be a description in the schema?

File: app/schemas/models/classroom.schema.coffee [hunk 25]
```diff
@@ -39,6 +39,19 @@ _.extend ClassroomSchema.properties,
...
+        type: 'array'
+        items: {
+          type: 'object'
+          additionalProperties: false
+          properties: {
+            levelOriginal: c.stringID()
+          }
+        }
+      }
+      first: {type: 'boolean'}   # is it the first level in the campaign
```
Comment (AndrewJakubowicz) [inline app/schemas/models/classroom.schema.coffee]: Comment vs adding it as a description in the schema?

File: app/views/play/VoyagerUnitMap.vue [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+            courseLevelsData[original].first = this.classroomLevelMap[original].first
+        }
+        // TODO: what if level was removed from campaign, and classroom level map doesnt contain positions, nextLevels, and first? Is it a valid scenario?
+        // If yes, then use the logic from CampaignView for setting the missing positions using the prev and next level positions
+        // Not sure how to set the .first and .nextLevels in that case.
+
+        return courseLevelsData
+      }
+      catch (err) {
+        return Promise.reject(err)
```
Comment (AndrewJakubowicz) [inline app/views/play/VoyagerUnitMap.vue]: If an await method fails, isn't `Promise.reject(err)` the default response?

File: app/views/play/VoyagerUnitMap.vue [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+      }
+    },
+    createLevelStatusMap() {
+      // Remove the level sessions for the levels played in another language - for the classroom version of unit map
+      // TODO: What if played in same language in another classroom
```
Comment (AndrewJakubowicz) [inline app/views/play/VoyagerUnitMap.vue]: Is this TODO something we are discussing solutions for? Or waiting for more info? Thanks!

File: app/views/play/VoyagerUnitMap.vue [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+            continue
+          const expectedLanguage = classroomLevel.primerLanguage || this.classroom.aceConfig.language
+          if (session.codeLanguage != expectedLanguage) {
+            this.levelSessions.splice(this.levelSessions.indexOf(session),1)
+          } 
+        }
+      }
+      this.levelStatusMap = utils.getLevelStatusMap(this.levelSessions)
+    },
+    determineNextLevel() { // set .next and .locked for this.levels
```
Comment (AndrewJakubowicz) [inline app/views/play/VoyagerUnitMap.vue]: In a second reading of this PR I can now see the differences between this `determineNextLevel` method and the `getNextLevelVoyager` method. I am also starting to see and experience the challenge of consolidating these various situations.

The `determineNextLevel` method has the following responsibilities:
 - Unlock levels for the user to play.
 - Also sets `.next` as a side effect.

Will the `.next` property *always* be in sync with what the `getNextLevelVoyager` returns?
An alternative is having a method to return levels to unlock. And then having a method to return next levels.
By decoupling these methods we can test them as separate concepts.

Finally in order to change state, updating the campaign view. We could layer these methods with different intentions and mutate the state based on their results.

The problem that I am trying to express is:
 - How can we ensure the various functions that are overlapping in functionality **always** agree?
 - Is there a chance that someone in the future updates this method and it is thus out of sync with `getNextLevelVoyager`?

File: test/app/core/voyagerUtils.spec.js [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+  return sessions // array of session objects
+}
+
+describe('voyager utilities', () => {
+  describe('findNextLevelsBySession returns an array of next level original ids for a given list of levels based on level sessions', () => {
+    beforeEach(() => {
+      me.set(factories.makeUser().attributes)
+    })
+
+    // c:completed, *:current level (not completed), n:not started (no level session)
```
Comment (AndrewJakubowicz) [inline test/app/core/voyagerUtils.spec.js]: I really like and appreciate the key.
Also helps see the test coverage. :clap:

File: test/app/views/play/VoyagerUnitMap.spec.js [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+    _.defer(done)
+  })
+
+  it('renders a vue instance', () => {
+    expect(unitMapClassroomWrapper.isVueInstance()).toBe(true)
+  })
+
+  it('shows the level dots for the classroom levels', () => {
+    expect(unitMapClassroomWrapper.find('.level-dot').exists()).toBe(true)
+    expect(unitMapClassroomWrapper.findAll('.level-dot').length).toBe(classroom.get('courses')[0].levels.length)
```
Comment (AndrewJakubowicz) [inline test/app/views/play/VoyagerUnitMap.spec.js]: I love this test. I just love it.

File: app/core/voyagerUtils.js (hunk 17 above)
Comment (shubhi1092) [reply app/core/voyagerUtils.js]: Thats a great suggestion. With the addition of 1FH capstone flow, we would most likely loop over `levels.nextLevels` to find the next level original (since the capstone will have multiple next levels, and only one will be returned based on the state of capstone level).
Will update it based on your suggestion when I will add the logic for capstone flow.

File: app/schemas/models/campaign.schema.coffee (hunk 23 above)
Comment (shubhi1092) [reply app/schemas/models/campaign.schema.coffee]: Good catch, I think I tried to use it while debugging an issue but it didnt help (not sure what it does entirely). Will remove. Thanks.

File: app/core/voyagerUtils.js (hunk 21 above)
Comment (shubhi1092) [reply app/core/voyagerUtils.js]: No particular reason, it's just using an object as a map.
However, I did re-check the usage of this method and it is being used to convert an array of levels to an object, in case its not already an object (https://github.com/codecombat/codecombat/pull/5281/files#diff-27dfda484676dea165b5a8540d8ec1d3R82). 
Returning an object from here will keep it in sync with the case when it is already an object.

File: app/lib/dynamicRequire.js (hunk 22 above)
Comment (shubhi1092) [reply app/lib/dynamicRequire.js]: Thats a good point, I am also going to rename voyager to ozaria at all places. How about updating this chunk name to be `ozariaPlay`?

File: app/schemas/models/classroom.schema.coffee (hunk 24 above)
Comment (shubhi1092) [reply app/schemas/models/classroom.schema.coffee]: Will add the specific descriptions.

File: app/views/play/VoyagerUnitMap.vue (hunk 27 above)
Comment (shubhi1092) [reply app/views/play/VoyagerUnitMap.vue]: This is an edge case that we didnt handle as of now. I am planning to clean up the TODO comments, and move them to asana/eng doc.

File: test/app/views/play/VoyagerUnitMap.spec.js (hunk 30 above)
Comment (shubhi1092) [reply test/app/views/play/VoyagerUnitMap.spec.js]: 😄

File: app/views/play/VoyagerUnitMap.vue (hunk 26 above)
Comment (shubhi1092) [reply app/views/play/VoyagerUnitMap.vue]: Yeah but if we don't catch it, wouldn't it result in an uncaught exception?

File: app/views/play/VoyagerUnitMap.vue (hunk 28 above)
Comment (shubhi1092) [reply app/views/play/VoyagerUnitMap.vue]: These are great questions.
I think that finding the next level in `findNextLevelsBySession` and then setting `.next` based on that here in `determineNextLevel` doesnt look like could get out of sync, since we set `.next` on whatever is returned from `findNextLevelsBySession`
However, I agree that setting `.locked` and setting `.next` logic can be decoupled a little more.
How about splitting `determineNextLevel` to something like this:
For unlocking levels:
```
for (let level in this.levels) {
          if (this.levelStatusMap[level] == 'started' || this.levelStatusMap[level] == 'complete' || this.levels[level].first || nextLevelOriginals.includes(level))
            this.levels[level].locked = false
          else
            this.levels[level].locked = true
        }
```
And for next levels:
```
for (let level in this.levels) {
          // there would only be one next level as per voyager v1 as of now
          if (nextLevelOriginals.includes(level)){
            this.levels[level].next = true
          }
        }
```
Not sure if I understand your questions right, and if this would address them?

Comment (AndrewJakubowicz) [reply app/views/play/VoyagerUnitMap.vue]: This makes the two intentions more readable and I like it for that reason.
Instead of two loops, they could be two separate map operations.
Your example is clearer as it keeps functionality separate with minimal runtime cost

Comment (AndrewJakubowicz) [reply app/views/play/VoyagerUnitMap.vue]: This makes the two intentions more readable and I like it for that reason.
Instead of two loops, they could be two separate map operations.
Your example is clearer as it keeps functionality separate with minimal runtime cost

File: app/views/play/VoyagerUnitMap.vue (hunk 26 above)
Comment (AndrewJakubowicz) [reply app/views/play/VoyagerUnitMap.vue]: I believe this will also result in an uncaught exception in promise. As we're not catching the "reject" here. We're unwrapping it and re-wrapping it. Not sure without testing. But async function returns a promise. And an error in that promise causes it to `reject`.

Comment (shubhi1092) [reply app/views/play/VoyagerUnitMap.vue]: Yeah I think you are right, and we are catching the error returned from here in the `created` method : https://github.com/codecombat/codecombat/pull/5281/files#diff-7629799da91ec72bc985d79e6a0e3fc8R70
Regarding async function always returning a rejected promise in case of error, is that true even if error occurs outside of await call, i.e. for example if this line causes an error: https://github.com/codecombat/codecombat/pull/5281/files#diff-7629799da91ec72bc985d79e6a0e3fc8R126

Comment (AndrewJakubowicz) [reply app/views/play/VoyagerUnitMap.vue]: I ran an experiment and you're right. Thanks for teaching me something new!
The warning about the unhandled error vanishes even if you re-wrap it as a rejection.
An alternative is adding a `.catch` to the function call. That also removes the warning.

Thank you!
